kernel/src/sched.rs

```diff
@@ -20,10 +28,60 @@ use crate::process::{self, Task};
 use crate::returncode::ReturnCode;
 use crate::syscall::{ContextSwitchReason, Syscall};
 
-/// The time a process is permitted to run before being pre-empted
-const KERNEL_TICK_DURATION_US: u32 = 10000;
 /// Skip re-scheduling a process if its quanta is nearly exhausted
-const MIN_QUANTA_THRESHOLD_US: u32 = 500;
+pub(crate) const MIN_QUANTA_THRESHOLD_US: u32 = 500;
+
+/// Trait which any scheduler must implement.
+pub trait Scheduler<C: Chip> {
+    /// Return the next process to run, and the timeslice length for that process.
+    /// Passing None instead of an AppId will lead to nothing being run.
+    /// Passing None instead of a timeslice will cause the process to be run cooperatively
+    fn next(&self) -> (Option<AppId>, Option<u32>);
+
+    /// Informs the scheduler of why the last process stopped executing, and how
+    /// long it executed for. Notably, execution_time_us will be `None`
+    /// if the the scheduler requested this process be run cooperatively.
+    fn result(&self, result: StoppedExecutingReason, execution_time_us: Option<u32>);
+
+    /// Tell the scheduler to execute kernel work such as interrupt bottom halves
+    /// and dynamic deferred calls. Most schedulers will use this default
+    /// implementation, but schedulers which at times wish to defer interrupt
+    /// handling will reimplement it.
+    ///
+    /// The reason for stubbing this out to allow schedulers to reimplement it
+    /// is that I (Hudson Ayers) believe that deferring bottom half
+    /// interrupt handling, or only servicing certain interrupts,
+    /// can be useful in helping userspace apps to meet
+    /// deadlines if a scheduler is given sufficient information about app
+    /// workloads. This stub could also potentially be useful for a power aware
+    /// scheduler which toggles between chip power modes based on preexisting knowledge
+    /// of process workloads.
+    unsafe fn execute_kernel_work(&self, chip: &C) {
+        chip.service_pending_interrupts();
+        DynamicDeferredCall::call_global_instance_while(|| !chip.has_pending_interrupts());
+    }
+
+    /// Ask the scheduler whether to take a break from executing userspace
+    /// processes to handle kernel tasks. Most schedulers will use this
+    /// default implementation, which always prioritizes kernel work, but
+    /// schedulers that wish to defer interrupt handling may reimplement it.
+    unsafe fn break_for_kernel_tasks(&self, kernel: &Kernel, chip: &C) -> bool {
+        chip.has_pending_interrupts()
+            || DynamicDeferredCall::global_instance_calls_pending().unwrap_or(false)
+            || kernel.processes_blocked()
+    }
```

Comment:
The interaction between `next()` and `break_for_kernel_tasks()` is a little surprising. Namely, I was expecting that if a scheduler returns `None` from `next()`, that seems like an indication for the chip to sleep. Instead, by default, the kernel will just continuously call `next()`.

Also, `break_for_kernel_tasks()` is actually `break_for_kernel_tasks_or_to_sleep()`.

What about removing `break_for_kernel_tasks()`, and adding the break to `next().map_or_else(|| break; ...`?

Reply:
The advantage of keeping `break_for_kernel_tasks()` is that all of the current schedulers use the default implementation, and it makes it very clear that the checks included in the default implementation should typically be used. 

I seem to recall you not liking when that code was re-implemented for each scheduler, which would be the case if it was moved into `next()`. But I agree the current interaction is confusing -- I am happy to take the approach of moving it into `next()` if you think that will be maintainable

Reply:
Ok I see, and yeah I didn't think about how that would force that duplicated logic into all of the `next()` calls. There are three clear decisions for the scheduler to make:

1. What process to run next, or none.
2. Whether to stop looking at processes and instead do kernel work.
3. Whether to continue to allow the current process to run.

And one more decision the kernel has to make:

4. Whether to go to sleep.

The three clear decisions each should be a function, like you have. I think what we need to establish is how does a scheduler decide if the chip should go to sleep, and how much control does it have, versus the kernel struct making some decisions to in theory help robustness.

Items (1) and (2) are related.

```
next()    | break_for_work()    | effect
-------------------------------------------
x         | true                | processes are not executed
some      | false               | that process is executed
none      | false               | ?
```

I believe right now the "?" is sit in a busy loop. I think instead that should be "try to sleep".

Reply:
In my proposed setup, if a scheduler returns `true` from `break_for_kernel_work()`, then we have to promise that `next()` will be called again before trying to sleep. That might be tricky to implement.